File: frontend/src/api/bidDraft.ts
import apiClient, { getAccessToken } from "@/api/client";

const API_BASE = (import.meta.env.VITE_API_BASE || "/api").replace(/\/$/, "");

export interface BidDraftOutlineNode {
  node_id: string;
  title: string;
  level: number;
  requirement?: string | null;
  article_count?: number;
  text_count?: number;
}

export interface BidDraftTask {
  id: string;
  project_id: string;
  tender_document_id: string | null;
  status: string;
  phase: string | null;
  analysis_result: Record<string, unknown> | null;
  outline: BidDraftOutlineNode[] | null;
  generation_options: Record<string, unknown> | null;
  summary: Record<string, unknown> | null;
  continue_of: string | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface BidDraftSectionMeta {
  node_id: string;
  title: string;
  status: string;
  word_count: number | null;
  attempts: number;
  error_message: string | null;
}

export interface BidDraftSectionContent {
  node_id: string;
  title: string;
  status: string;
  content: string | null;
  word_count: number | null;
}

export interface BidDraftAssembled {
  task_id: string;
  status: string;
  content: string | null;
  word_count: number | null;
  section_total: number;
  section_generated: number;
  section_failed: number;
}

export async function createBidDraftTask(payload: {
  project_id: string;
  tender_document_id: string;
  outline?: BidDraftOutlineNode[];
  analysis?: Record<string, unknown> | null;
  options?: Record<string, unknown> | null;
}) {
  const response = await apiClient.post<BidDraftTask>("/bid-draft/tasks", payload);
  return response.data;
}

export async function getBidDraftTask(taskId: string) {
  const response = await apiClient.get<BidDraftTask>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}`,
  );
  return response.data;
}

export async function cancelBidDraftTask(taskId: string) {
  const response = await apiClient.post<BidDraftTask>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}/cancel`,
  );
  return response.data;
}

export async function listBidDraftSections(taskId: string) {
  const response = await apiClient.get<BidDraftSectionMeta[]>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}/sections`,
  );
  return response.data;
}

export async function getBidDraftSectionContent(taskId: string, nodeId: string) {
  const response = await apiClient.get<BidDraftSectionContent>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}/sections/${encodeURIComponent(nodeId)}`,
  );
  return response.data;
}

export async function getBidDraftAssembled(taskId: string) {
  const response = await apiClient.get<BidDraftAssembled>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}/assembled`,
  );
  return response.data;
}

export async function regenerateBidDraftSection(taskId: string, nodeId: string) {
  const response = await apiClient.post<BidDraftTask>(
    `/bid-draft/tasks/${encodeURIComponent(taskId)}/sections/${encodeURIComponent(nodeId)}/regenerate`,
  );
  return response.data;
}

export function bidDraftStreamUrl(taskId: string) {
  return `${API_BASE}/bid-draft/tasks/${encodeURIComponent(taskId)}/stream`;
}

export function bidDraftToken() {
  return getAccessToken();
}

File: backend/schemas/bid_draft.py
"""Pydantic schemas for bid draft generation tasks."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

OUTLINE_MAX_NODES = 60


class OutlineNode(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    level: int = Field(default=1, ge=1, le=6)
    requirement: str | None = Field(default=None, max_length=2_000)
    article_count: int = Field(default=2, ge=1, le=8)
    text_count: int = Field(default=400, ge=100, le=3_000)


class BidDraftOptions(BaseModel):
    only_sections: list[str] | None = Field(default=None, max_length=OUTLINE_MAX_NODES)
    outline_hint: str | None = Field(default=None, max_length=2_000)


class BidDraftTaskCreate(BaseModel):
    project_id: str = Field(min_length=1, max_length=36)
    tender_document_id: str = Field(min_length=1, max_length=36)
    outline: list[OutlineNode] | None = Field(default=None, max_length=OUTLINE_MAX_NODES)
    analysis: dict[str, Any] | None = None
    options: BidDraftOptions | None = None


class BidDraftTaskResponse(BaseModel):
    id: str
    project_id: str
    tender_document_id: str | None
    status: str
    phase: str | None
    analysis_result: dict[str, Any] | None
    outline: list[Any] | None
    generation_options: dict[str, Any] | None
    summary: dict[str, Any] | None
    continue_of: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BidDraftSectionResponse(BaseModel):
    node_id: str
    title: str
    status: str
    word_count: int | None
    attempts: int
    error_message: str | None

    model_config = ConfigDict(from_attributes=True)


class BidDraftSectionContentResponse(BaseModel):
    node_id: str
    title: str
    status: str
    content: str | None
    word_count: int | None


class BidDraftAssembledResponse(BaseModel):
    task_id: str
    status: str
    content: str | None
    word_count: int | None
    section_total: int
    section_generated: int
    section_failed: int
